Copter: Separate max ascent and descent speeds

ChrisBird opened the pull request:
In reference to issue #6879 

This is for the Copter components only, other pull requests will be generated for Sub and hopefully Plane (quadplane).  Change should stand on its own (no library changes, etc)

Note that I've used parameter names that people seemed okay those replying in the issue, if they need changing let me know.

Tested on SITL.

Apologies on the commits (only the last commit is relevant to the copter change but I couldn't get rid of the others - a file got reverted to an old version and committed by accident.  I've reverted it back - let me know if I need to fix this up more than I've done to get the pull request through) - still learning git and learning how to fix my mistakes.  Only files updated and changed are the Copter files so I think I've managed to fix my mistake.

Pedals2Paddles commented:
Nice job! I would backup the files you need and do a git reset on your branch. And then rebase to master. Then add your files again and do the commit. Then force push the branch which will update this PR. That will fix the mess of commits. The pile of merges and reverts may actually make Francisco throw up on his keyboard, and the won't be accepted as it is.

OXINARF commented on ArduCopter/Parameters.h:
```diff
@@ -368,6 +368,12 @@ class Parameters {
+        // Pilot_Velocity Extras - 260
+        //
+        k_param_pilot_velocity_z_max_up = 260,
+        k_param_pilot_velocity_z_max_dn = 261,
```
This needs to be verified, but I think we can't use these numbers, that's why we already have another group of parameters.

ChrisBird replied:
Hi @OXINARF, I must be misunderstanding something?  I'm still learning the codebase (and my c++ skills are a little rusty) so apologies if I'm making basic mistakes.  I gave the new ones a new block and gave them values not in the enum list?  

Thanks for the above git suggestions, I'll give it a go when I get home.

rmackay9 commented:
In general looking pretty good.

We should continue to use the old enum location even though we are renaming the parameter to "PILOT_SPEED_UP" or everyone will lose whatever their current value is (although many do not change it).

I think we should do one of these two things:

- have the "down" parameter default as 0 and if it's zero, use the "up" parameter value.  That would keep the behaviour as it is now but allow for an optional different down speed.
- copy the up value to the down value as part of the parameter conversion.  The conversion code is near the bottom of Parameters.cpp. https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/Parameters.cpp#L1046

ChrisBird commented:
I'll make some changes tonight (fixing up my commits) and I'll add @rmackay9 2nd suggestion as it would be cleaner.  I knew it could do the conversion but never knew how, looks straight forwards.

ChrisBird commented:
@rmackay9 @OXINARF I've updated it as per the recommendations above.  Since the parameter for dn was on it own now, I moved it to 256 just passed the reserved values.

I think I even managed to get it to fit into the one commit too, starting to understand how git works now.

Let me know if anything else needs changing.

ChrisBird commented:
Ah i see what you mean now. I'll update it to use the other block.  I'll do the first option. I'll get an update out today.

ChrisBird commented:
@OXINARF, I think I've resolved all the issues identified.  Given we've run out of room in the top level params it might be worthwhile updating the How to on the dev page.  If I get a chance I'll give updating the doco a go - seems its more like what we have in the adding a param to a library.

Reason I initially did it is due to this comment and the how to add a new param page:
        // the k_param_* space is 9-bits in size
        // 511: reserved

I assumed I was good to use all 9 bits, I did think it was odd at the time and didn't fully understand the earlier parameter comment you made, the 2nd one made sense to me and hopefully I've done it correctly.  Seems to work on my SITL tests.

Maybe we should change that first line to:
        // the k_param_* space is 9-bits in size.  However only 8 bits is accessible

Anyway, I've updated the code so it should be using G2 parameters instead and I believe I've fixed the conversion issue you raised.

I'll get there its only my 2nd PR, still learning.  Thanks for your help!

ChrisBird commented:
Hi @rmackay9, let me know if it needs any further work?

OXINARF commented:
@ChrisBird Unfortunately I was late to the dev call where this was discussed so I'm not completely sure what @rmackay9  said there, but from reading the minutes of the meeting, it looks like he wants to have the other option implemented instead (keep the same parameter and add a new one with default value of 0, which means *use existing parameter*).

ChrisBird commented:
@OXINARF I'll make the change over the weekend.

ChrisBird commented:
Hi @OXINARF, I think I've done something wrong on the rebase to get all the commits above there.  I've put a request out on Gitter to seek help but if now one gets back to me can you advise what I should do?

Should I be using the fixup option?  Or is some other course of action recommended.  Apologies for this, it seems Git is still kicking my butt.  I'll get it eventually.....

Also I'm confused why the conflict messages, as I resolved the conflicts on my end (winch was using the same parameter number as me orginally), so I'm not sure why they have appeared here.

OXINARF commented:
@ChrisBird Waiting for you in Gitter 🙂 No, don't use fixup, you need to rebase.

ChrisBird commented:
@rmackay9 Can you let me know if the changes I've made are in line with what was discussed on the dev call?  I've only got the minutes to work off and I'd like to get this one finalised.  The issue I mentioned above, @OXINARF was able to help me resolve on the day.

OXINARF commented on ArduCopter/Attitude.cpp:
```diff
@@ -307,3 +307,12 @@ void Copter::rotate_body_frame_to_NE(float &x, float &y)
 }
+
+// It will return the PILOT_SPEED_DN value if non zero, otherwise if zero it returns the PILOT_SPEED_UP value.
+uint16_t Copter::get_pilot_z_max_dn_velz()
```
Nitpicking: I think this was better named `get_pilot_velocity_z_max_dn`.

ChrisBird replied:
It makes sense, I've changed it.

OXINARF commented on ArduCopter/control_sport.cpp:
```diff
@@ -68,7 +68,7 @@ void Copter::sport_run()
     // get pilot desired climb rate
     float target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
-    target_climb_rate = constrain_float(target_climb_rate, -g.pilot_velocity_z_max, g.pilot_velocity_z_max);
+    target_climb_rate = constrain_float(target_climb_rate, -g2.pilot_velocity_z_max_dn, g.pilot_velocity_z_max_up);
```
It's not calling the method.

ChrisBird replied:
Hmmm, annoyed at myself on this one.  I've fixed it up in the latest commit.

OXINARF commented on ArduCopter/config.h:
```diff
@@ -544,12 +544,16 @@
 #endif
+#ifndef PILOT_SPEED_DEF
+ # define PILOT_SPEED_DEF    250     // maximum vertical velocity in cm/s
+#endif
```
I think we don't need to change the macro name, but if you want to do that, change the old macro name, don't add a new one and leave the old, unused, one.

ChrisBird replied:
I've reverted to using the old name, removed the PILOT_SPEED_DEF and removed the comment deprecating the original name.

OXINARF commented on ArduCopter/Attitude.cpp:
```diff
@@ -307,3 +307,12 @@ void Copter::rotate_body_frame_to_NE(float &x, float &y)
+// It will return the PILOT_SPEED_DN value if non zero, otherwise if zero it returns the PILOT_SPEED_UP value.
+uint16_t Copter::get_pilot_z_max_dn_velz()
+{
+    if(g2.pilot_velocity_z_max_dn == 0)
```
Style rules: space after *if* and you need braces

ChrisBird replied:
Updated in latest commit

OXINARF commented on ArduCopter/Attitude.cpp:
```diff
@@ -307,3 +307,12 @@ void Copter::rotate_body_frame_to_NE(float &x, float &y)
+    if(g2.pilot_velocity_z_max_dn == 0)
+        return g.pilot_velocity_z_max_up;
+    else
+        return g2.pilot_velocity_z_max_dn;
```
It would probably be a good idea to put an *abs* around this to prevent anyone inserting a negative value.

ChrisBird replied:
Agreed, updated in latest commit

ChrisBird commented:
@rmackay9 I decided to do the changes now.  They are done however it looks like the travis ci sitltest failed (in rover code, looks like others have had the same issue from what I could see), so once that is fixed it should pass the sitltest.

Pedals2Paddles commented:
Looks like all checks have passed now.

rmackay9 commented:
Merged, thanks!

Pedals2Paddles commented:
I used this today and it was magical! Little things like this make a big difference when this little thing has been an annoyance for years.

I set a max descent rate that will be smooth, just on the edge of where it starts to wobble in it's own wake turbulence.  This made the video I was shooting today so much easier to manage and has an immediate effect on quality.  I no longer need to micro manage my descent rate and I can focus on the shot I'm trying to get.

Thank you!

ChrisBird commented:
No probs, glad you've been able to test it out.  It'll be a few weeks before i can fly it on a real multi.  I'll be going through the issues list and working on some of the older simple items to help me learn the code.
